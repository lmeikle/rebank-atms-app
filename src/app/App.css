.App {
  max-width: 400px;
  margin-left: auto;
  margin-right: auto;
  text-align: center;
  font-family: 'Titillium Web', sans-serif;
  color: #002742;
}

a{text-decoration:none}
a{color:#1e70bf}
a:hover{text-decoration:none}
a:hover{color:#ED3773}

.AppTopBar {
  height: 70px;
  background-color: #FFFFFF;
}

.AppTorBarLogo {
  margin-top: 20px;
  width: 40%;
}
