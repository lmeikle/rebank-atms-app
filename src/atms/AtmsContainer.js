import React, { Component } from 'react';
import { fetchAtmData } from './AtmsAPI';
import AtmsComponent from './AtmsComponent';
import LoadingComponent from '../loading/LoadingComponent';
import calculateDistance from '../utils/calculateDistance';
import getGeolocation from '../utils/getGeolocation';

/**
 * Renders the list of nearest atms for the selected bank
 */
class AtmsContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: null,
      geolocation: null,
      atms: null,
      error: null
    };
  }

  static get NEAREST_ATM_QUANTITY() {
    return 10;
  }

  componentDidMount() {
    if (this.props.location.state) {
      let { name, url } = this.props.location.state;
      if (url) {
        // reset some of the state
        this.setState({
          name,
          atms: null,
          error: null
        });

        getGeolocation()
          .then(geolocation => this.setState({ geolocation }))
          .catch(error => this.setState({ error }));

        fetchAtmData(url)
          .then(atms => this.setState({ atms }))
          .catch(error => this.setState({ error: error.toString() }));

        return;
      }
    }

    // else go the default screen
    this.props.history.replace('/');
  }

  /**
	 * Commented out as there is no performance issue at present.
	 *
	 * As the filtering of ATM's to display is performed within the render function (because maybe we will
	 * update geolocation in the future or even change the filter criteria) I want to be sure to only run the
	 * update it if the state has actually changed
	 *
	shouldComponentUpdate(nextProps, nextState)
	{
		return (
			nextState.geolocation !== this.state.geolocation ||
			nextState.atms !== this.state.atms ||
			nextState.error !== this.state.error
		);
	}*/

  static findNearestATMs(geolocation, atms, size) {
    if (atms && geolocation) {
      let atmsWithDistance = atms.map(atm => ({
        ...atm,
        distance: calculateDistance(
          geolocation.coords.latitude,
          geolocation.coords.longitude,
          parseFloat(atm.coords.Latitude, 10),
          parseFloat(atm.coords.Longitude, 10)
        )
      }));

      atmsWithDistance.sort((a, b) => {
        if (a.distance < b.distance) return -1;
        if (a.distance > b.distance) return 1;
        return 0;
      });

      return atmsWithDistance.slice(0, size);
    }

    return null;
  }

  render() {
    const { name, geolocation, atms, error } = this.state;

    if (error) {
      return (
        <div className="errorMessage">
          Failed to find nearest {name} ATM's due to: {error.toString()}
        </div>
      );
    }

    // filter the atm data
    let nearestAtms = AtmsContainer.findNearestATMs(geolocation, atms, AtmsContainer.NEAREST_ATM_QUANTITY);
    if (!nearestAtms) {
      return <LoadingComponent />;
    }

    return (
      <div>
        <div>
          The {AtmsContainer.NEAREST_ATM_QUANTITY} nearest {name} ATM's are:
        </div>
        {nearestAtms.map(atm => <AtmsComponent key={atm.identification} {...atm} />)}
      </div>
    );
  }
}

export default AtmsContainer;
