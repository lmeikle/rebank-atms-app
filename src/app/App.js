import React, { Component } from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import AtmsContainer from '../atms/AtmsContainer';
import BanksContainer from '../banks/BanksContainer';
import logo from './logo.svg';
import './App.css';

class App extends Component {
	render()
	{
		return (
			<div className="app">
				<div className="app-topbar">
					<img src={logo} className="app-topbar-logo" alt="logo"/>
				</div>
				<Router>
					<Switch>
						<Route path="/atms" component={AtmsContainer}/>
						<Route path="/" component={BanksContainer}/>
					</Switch>
				</Router>
			</div>
		);
	}
}

export default App;
