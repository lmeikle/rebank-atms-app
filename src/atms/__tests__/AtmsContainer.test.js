import React from 'react';
import { mount, shallow } from 'enzyme';
import AtmsContainer from '../AtmsContainer';
import App from '../../app/App';

const props = {
  location: {
    state: {
      name: 'Barclays Bank',
      url: 'www.example.com'
    }
  },
  history: ''
};

const initialState = {
  geolocation: null,
  atms: null,
  error: null
};

const readyState = {
  geolocation: { coords: { latitude: 123, longitude: 456 } },
  atms: [
    {
      identification: '456',
      name: 'Barclays Bank',
      countryCode: 'UK',
      address: '10 Watford High Street',
      coords: { Latitude: 234, Longitude: 567 },
      distance: 5.23,
      numOfAtms: 1
    },
    {
      identification: '123',
      name: 'Barclays Bank',
      countryCode: 'UK',
      address: '10 Camden High Street',
      coords: { Latitude: 123, Longitude: 456 },
      distance: 1.23,
      numOfAtms: 3
    },
    {
      identification: '789',
      name: 'Barclays Bank',
      countryCode: 'UK',
      address: '10 Chesham High Street',
      coords: { Latitude: 345, Longitude: 567 },
      distance: 10.23,
      numOfAtms: 1
    }
  ],
  error: null
};

const errorState = {
  geolocation: { coords: { latitude: 123, longitude: 456 } },
  atms: null,
  error: 'There was an error'
};

const atmsContainer = shallow(<AtmsContainer {...props} />);

describe('AtmsContainer', () => {
  test('renders correctly', () => {
    expect(atmsContainer).toMatchSnapshot();
  });

  test('initializes the `state` correctly', () => {
    expect(atmsContainer.state()).toEqual(initialState);
  });

  test('renders a list of AtmsComponents', () => {
    atmsContainer.setState(readyState);
    expect(atmsContainer.find('AtmsComponent').exists()).toBe(true);
  });

  test('given an arrry of atms find the nearest atms', () => {
    expect(AtmsContainer.findNearestATMs(readyState.geolocation, readyState.atms, 1)[0].address).toBe(
      '10 Camden High Street'
    );
  });

  test('displays a error message', () => {
    atmsContainer.setState(errorState);
    expect(atmsContainer.find('.errorMessage').exists()).toBe(true);
  });
});
