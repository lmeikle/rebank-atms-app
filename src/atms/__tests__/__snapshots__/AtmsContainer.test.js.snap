// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtmsContainer renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AtmsContainer
    history=""
    location={
        Object {
            "state": Object {
              "name": "Barclays Bank",
              "url": "www.example.com",
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
